### PR TITLE
Aos 2 fixes efif

### PR DESCRIPTION
**Pull request by efifogel**
## Summary of Changes

This PR contains a bunch of changes that address issues found while I worked on the 2nd edition of the "CGAL Arrangement Book". In the Book we offer the code of several programs in Python. The code is based on the recent Python bindings. Several improvements to the code were necessary to support what I refer to "layered extension". For example, `Arrangement_on_surface_with_history_2` derives from `Arrangement_on_surface_2` and requires extensions to the DCEL records. `Envelope_diagram_on_surface_2` also derives from `Arrangement_on_surface_2` and requires extensions to the DCEL records. The improvements applied enables having  `Envelope_diagram_on_surface_with_history_2` (which, naturally, requires both sets of extensions). I also seized the opportunity and made the code look more modern, for example, replaced a bunch of "typedef" statements with "using" statements. 

1. Envelope_2
* Introduced move constructor and move assignment operator for the main class `Envelope_diagram_1`.

2. [Arrangement_on_surface_2](https://www.cs.tau.ac.il/~efif/doc_output8/Arrangement_on_surface_2/)
* Introduced `Arr_dcel`, which essentially replaces the former `Arr_default_dcel`. Backward compatibility was maintained by the (re) introduction of the alias template `Arr_default_dcel`. `Arr_dcel`, as opposed to the former `Arr_default_dcel` is parameterized by (in addition to the geometry traits) Vertex, Halfedge, and Face template parameters, and they have default type values. All this enables the layered extension of DCEL records; see [Arr_dcel](https://www.cs.tau.ac.il/~efif/doc_output8/Arrangement_on_surface_2/classCGAL_1_1Arr__dcel.html)
* Replaced some tex code, which uses pstricks, that is used to generate figures in the manual with code that uses tikz. (All figures in the book use tikz.)
* Fixed some typos and text in the manual, for example, replaced the phrase "substituted with" with "substituted by" when referring to a template parameter being substituted by a type when the template is instantiated. This matches the text in the book.
* Made `Curve_halfedges` nested in  `Arrangement_on_surface_with_history_2` public, as we enable iterating over such types (so it really makes no sense not to have it public). I haven't reflected it in the manual. Publishing this, I believe, requires a small feature, and I'll do it separately.

3. [Envelope_3](https://www.cs.tau.ac.il/~efif/doc_output8/Envelope_3/)
* Refreshed the example programs, e.g., used 'auto' when it made sense and replaced "typedef" statements with "using" statements. 
* Eliminated the construction of "local" kernel objects and local traits objects in the various trait classes, namely, `Env_triangle_traits_3`, `Env_plane_traits_3`, and `Env_sphere_traits_3`. (Some local objects still remain; however, eliminating them will require breaking backward compatibility, so it is left for a future PR.)
* Passed the base class of `Env_plane_traits_3` as a template parameter with a default value (being the 2D arrangement linear traits); see [Env_plane_traits_3](https://www.cs.tau.ac.il/~efif/doc_output8/Envelope_3/classCGAL_1_1Env__plane__traits__3.html). Similarly, passed the base class of `Env_triangle_traits_3` as a template parameter with a default value (being the 2D arrangement segment traits); see [Env_triangle_traits_3](https://www.cs.tau.ac.il/~efif/doc_output8/Envelope_3/classCGAL_1_1Env__triangle__traits__3.html). See also [EnvelopeTraits_3 concept](https://www.cs.tau.ac.il/~efif/doc_output8/Envelope_3/classEnvelopeTraits__3.html) . This chanhes is required for "layered extension".
* Made all members of the cell-extension records of the DCEL used by the envelope diagram unique (in particular, added '    env_'); this avoids name-collisions with other extensions required by other packages.
* Added template parameters Vertex, Halfedge, and Face with default values to `Envelope_pm_dcel`. This is similar to `Arr_dcel`; see above. BW,  `Envelope_pm_dcel` is not published, so there is no change to the manual. In fact, I don't see a reason not to publish it, but this will come in a future PR.
* Added missing 'const' in some signatures

## Release Management

* Affected package(s): Arrangement_on_surface_2, Envelope_2, and Envelope_3
* Issue(s) solved (if any): NA
* Feature/Small Feature (if any): NA
* Link to compiled documentation (obligatory for small feature) [new doc](https://www.cs.tau.ac.il/~efif/doc_output8)
* License and copyright ownership: TAU

### CODE REVIEW

File: Envelope_3/include/CGAL/Env_sphere_traits_3.h [hunk 1]
```diff
@@ -895,8 +971,12 @@ class Env_sphere_traits_3 : public ConicTraits_2 {
...
     // and in general we have:
     //    -m*x + n*y + (m*x0 -n*y0) = 0 (with integer coordinates)
-    const Rational r = cv.r(), s = cv.s(), t = cv.t(),
-               u = cv.u(), v = cv.v(), w = cv.w();
+    const Rational& r = cv.r();
+    const Rational& s = cv.s();
+    const Rational& t = cv.t();
+    const Rational& u = cv.u();
+    const Rational& v = cv.v();
+    const Rational& w = cv.w();
```
Comment (sloriot) [inline Envelope_3/include/CGAL/Env_sphere_traits_3.h:979]: unused variable `w`

File: Envelope_3/include/CGAL/Env_plane_traits_3.h [hunk 2]
```diff
@@ -201,35 +183,63 @@ class Env_plane_traits_3 : public Arr_linear_traits_2<Kernel_>
...
 
       // our line is a3*x + b3*y + c3 = 0
       // it is assumed that the planes intersect over this line
       const Line_2& line = cv.supp_line();
-      const FT& a3 = line.a(),
-        b3 = line.b(),
-        c3 = line.c();
+      const FT& a3 = line.a();
+      const FT& b3 = line.b();
+      const FT& c3 = line.c();
```
Comment (sloriot) [inline Envelope_3/include/CGAL/Env_plane_traits_3.h:242]: Warning: unused variable c3

**Comment by efifogel** [comment]
Fixed both
   ____  _        ____             _
  /_____/_) o    /__________  __  //
 (____ (   (    (    (_/ (_/-(-'_(/
                         _/

**Comment by sloriot** [comment]
Successfully tested in CGA-6.0-Ic-173

**Pull request merged**